### PR TITLE
:seedling: Make load balancer first-party package in CAPD

### PR DESCRIPTION
**Pull request by killianmuldoon**
Moves the loadbalancer package from third_part to first part in the Docker provider. This lets us update and possibly make changes to the load balancer without needing to sync with the kind implementation.

This version still uses the kindest/haproxy image so it's still pretty tightly coupled with Kind.

Note this should fix the issue described in #7344, but those changes are also available in the Kind loadbalancer library now.

### CODE REVIEW

File: test/infrastructure/docker/internal/loadbalancer/doc.go [hunk 1]
```diff
@@ -14,5 +14,6 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-// Package loadbalancer contains external loadbalancer related constants and configuration
+// Package loadbalancer contains external loadbalancer related constants and configuration.
+// This package was originally forked from the kind load balancer at https://github.com/kubernetes-sigs/kind/tree/v0.7.0/pkg/cluster/internal/loadbalancer
```
Comment (killianmuldoon) [inline test/infrastructure/docker/internal/loadbalancer/doc.go:18]: Not sure if this is the correct level of attribution, open to different opinions here for sure.

File: test/infrastructure/docker/internal/loadbalancer/config.go [hunk 2]
```diff
@@ -23,19 +23,25 @@ import (
...
 
-// DefaultConfigTemplate is the loadbalancer config template
-const DefaultConfigTemplate = `# generated by kind
+// ConfigTemplate is the loadbalancer config template.
+const ConfigTemplate = `# generated by kind
 global
   log /dev/log local0
   log /dev/log local1 notice
   daemon
+  # limit memory usage to approximately 18 MB
```
Comment (dlipovetsky) [inline test/infrastructure/docker/internal/loadbalancer/config.go:39]: I would appreciate a comment that explains where this limit came from, because I think it's far from self-explanatory :laughing: 

```suggestion
  # limit memory usage to approximately 18 MB
  # (see https://github.com/kubernetes-sigs/kind/pull/3115)
```

Comment (killianmuldoon) [reply test/infrastructure/docker/internal/loadbalancer/config.go:39]: Will add - thanks for getting that merged btw! Very useful fix :smile:

File: docs/book/src/developer/providers/v1.3-to-v1.4.md [hunk 3]
```diff
@@ -70,7 +70,8 @@ maintainers of providers and consumers of our Go API.
   For more information, please see: https://github.com/kubernetes/enhancements/issues/2845
 - A new `KCPRemediationSpec` test has been added providing better test coverage for KCP remediation most common use cases. As a consequence `MachineRemediationSpec` now only tests remediation of
   worker machines (NOTE: we plan to improve this test as well in a future iteration).
-
+- Package `test/infrastructure/docker/internal/third_party/forked/loadbalancer` has been moved to `test/infrastructure/docker/internal/loadbalancer` to allow it to diverge from the upstream Kind package.
```
Comment (sbueringer) [inline docs/book/src/developer/providers/v1.3-to-v1.4.md:73]: nit/Q: Is this relevant to provider authors?

**Comment by sbueringer** [comment]
Q: How similar is this package now to the upstream kind package? (related: how equal was it before this PR) Is the maintenance now entirely on us?

**Comment by killianmuldoon** [comment]
>Q: How similar is this package now to the upstream kind package? (related: how was it before this PR) Is the maintenance now entirely on us?

It's pretty similar on purpose - I fixed up some linter findings but that's it. Before it was pretty out of date. Because this was a `third_party` package maintenance was always on us, but we never really did anything as it continued to work. This change is mostly about making that maintenance explicit.

File: docs/book/src/developer/providers/v1.3-to-v1.4.md (hunk 3 above)
Comment (killianmuldoon) [reply docs/book/src/developer/providers/v1.3-to-v1.4.md:73]: As it's still internal maybe not - but because it's a slightly strange sort of change - third_party -> first_party - I wanted to highlight it somewhere.

Comment (sbueringer) [reply docs/book/src/developer/providers/v1.3-to-v1.4.md:73]: No objection

File: test/infrastructure/docker/internal/loadbalancer/doc.go (hunk 1 above)
Comment (sbueringer) [reply test/infrastructure/docker/internal/loadbalancer/doc.go:18]: Probably it's fine given that we copy from kubernetes-sigs to kubernetes-sigs.

If you want to make sure I would ask Ben.

File: test/infrastructure/docker/internal/loadbalancer/const.go [hunk 4]
```diff
@@ -16,14 +16,16 @@ limitations under the License.
 
 package loadbalancer
 
-// Image defines the loadbalancer image name
-const Image = "haproxy"
+const (
```
Comment (richardcase) [inline test/infrastructure/docker/internal/loadbalancer/const.go:19]: We could probably also change to use the "official" haproxy image, so

```go
Image = "haproxy-alpine"
DefaultImageRepository = "haproxytech"
DefaultImageTag        = "2.8"
```

Comment (killianmuldoon) [reply test/infrastructure/docker/internal/loadbalancer/const.go:19]: The issue (same as the issue for the Kind repo) is the official image doesn't have a config file attached which means it won't start. We could definitely workaround this by setting up the image in the code, but AFAIU CAPI needs to make some changes either way.

Comment (richardcase) [reply test/infrastructure/docker/internal/loadbalancer/const.go:19]: We used the official image for the sample provider used in the kubecon na tutorial: https://github.com/capi-samples/cluster-api-provider-docker/tree/main/pkg/loadbalancer

I don't overly remember us having to change anything in the `loadbalancer` package but i can check.

**Comment by killianmuldoon** [comment]
/hold

Waiting for input on https://github.com/kubernetes-sigs/cluster-api/pull/8246#discussion_r1129574113

File: test/infrastructure/docker/internal/loadbalancer/const.go (hunk 4 above)
Comment (sbueringer) [reply test/infrastructure/docker/internal/loadbalancer/const.go:19]: Let's please consider that the more we move away from what kind is doing (probably) the more maintenance effort we have

File: test/infrastructure/docker/internal/loadbalancer/doc.go (hunk 1 above)
Comment (BenTheElder) [reply test/infrastructure/docker/internal/loadbalancer/doc.go:18]: Kubernetes has done this sort of thing before, including in KIND!

IMHO we have a nice shortcut here in that these projects are both under the same license and "owner" (the project / contributors) so it's fine to just copy. I recommend attribution comments like this both for historical context and to acknowledge that another project originally wrote them.

So yes, this is fine to me. For prior art see also kubeadm/kinder tool 🙂

File: test/infrastructure/docker/internal/loadbalancer/const.go (hunk 4 above)
Comment (killianmuldoon) [reply test/infrastructure/docker/internal/loadbalancer/const.go:19]: Thanks for the pointer @richardcase - this is working. Looks like it's the 'global' haproxy image which doesn't have a config loaded. The haproxytech image has a default config which lets it run initially.

>Let's please consider that the more we move away from what kind is doing (probably) the more maintenance effort we have. After all CAPD clusters are still very closely modeled after kind clusters and there are dependencies.

I'm fine with pushing changing the image to a follow up - but I don't think there's additional maintenance effort here compared to using a third_party folder. The change is that this part of CAPD wasn't really maintained at all previously.

File: test/infrastructure/docker/internal/loadbalancer/doc.go (hunk 1 above)
Comment (sbueringer) [reply test/infrastructure/docker/internal/loadbalancer/doc.go:18]: Sounds good, thx for chiming in!

File: test/infrastructure/docker/internal/loadbalancer/const.go (hunk 4 above)
Comment (sbueringer) [reply test/infrastructure/docker/internal/loadbalancer/const.go:19]: Up to you. I didn't look into it in detail, but if you're sure that there are no dependencies to the kind haproxy I'm fine with switching to the "upstream" one. I just know that we have a lot of dependencies to the kindest/node image and even make sure that we start the node image exactly the same way as kind does.

If we don't have stuff like this for the lb image, I agree, it shouldn't matter

Comment (sbueringer) [reply test/infrastructure/docker/internal/loadbalancer/const.go:19]: The comment in the kind repo seems to match what Killian was writing above
> We cannot merely use the upstream haproxy image as haproxy will exit without a minimal config, so we introduce one that will list on the intended port and hot reload it at runtime with the actual desired config.
> https://github.com/kubernetes-sigs/kind/tree/main/images/haproxy

Comment (killianmuldoon) [reply test/infrastructure/docker/internal/loadbalancer/const.go:19]: Exactly - this image exists with the `_/haproxy` image, but not the `haproxytext/haproxy` image. 

I'm 100 percent fine with sticking with the haproxy image from kind for now btw - I don't see any motivation to change this. That said, making this directory first party will allow changes of that kind in future if someone is interested.

For now let's not change it.

Comment (sbueringer) [reply test/infrastructure/docker/internal/loadbalancer/const.go:19]: Sounds good!

**Comment by sbueringer** [comment]
/hold cancel
given answer above

**Review by fabriziopandini — COMMENTED** [review]
/lgtm
/approve
